src/main/java/algorithms/curated170/medium/closestleafinabinarytree/ClosestLeafInABinaryTreeGraphBFS.java

```diff
@@ -0,0 +1,82 @@
+package algorithms.curated170.medium.closestleafinabinarytree;
+
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Deque;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import algorithms.datastructures.TreeNode;
+
+public class ClosestLeafInABinaryTreeGraphBFS {
+
+  Map<TreeNode, List<TreeNode>> graph;
+  int k;
+  TreeNode nodeK = null;
+
+  public int findClosestLeaf(TreeNode root, int k) {
+    if (root.left == null && root.right == null) {
+      return k;
+    }
+
+    this.k = k;
+    graph = new HashMap<>();
+    createGraphWithDFS(root, null);
+
+    return shortestPathBFS(root);
+  }
+
+  private int shortestPathBFS(TreeNode root) {
+    Deque<TreeNode> queue = new ArrayDeque<>();
+    boolean[] seen = new boolean[1001];
+
+    queue.offer(nodeK);
+    seen[k] = true;
+
+    while (!queue.isEmpty()) {
+      TreeNode node = queue.poll();
+      if (node != null) {
+        if (graph.get(node).size() <= 1 && node != root) {
+          return node.val;
+        }
+        for (TreeNode adjacentNode : graph.get(node)) {
+          if (!seen[adjacentNode.val]) {
+            seen[adjacentNode.val] = true;
+            queue.add(adjacentNode);
+          }
+        }
+      }
+    }
+
+    throw null;
+  }
+
+  public void createGraphWithDFS(TreeNode node, TreeNode parent) {
+    if (node != null) {
+      if (node.val == k) {
+        nodeK = node;
+      }
+      graph.put(node, new ArrayList<>());
+      if (parent != null) {
+        graph.get(node).add(parent);
+        graph.get(parent).add(node);
+      }
+
+      createGraphWithDFS(node.left, node);
+      createGraphWithDFS(node.right, node);
+    }
+  }
+
+  public static void main(String[] args) {
+    var solution = new ClosestLeafInABinaryTreeGraphBFS();
+
+    TreeNode tn1 = new TreeNode(1);
+    TreeNode t0 = new TreeNode(2, null, tn1);
+    TreeNode t1 = new TreeNode(5);
+    TreeNode t2 = new TreeNode(4, null, t1);
+    TreeNode t3 = new TreeNode(3, t0, t2);
+    TreeNode root = new TreeNode(9, null, t3);
+
+  }
+}
```

Comment:
The code has no problem I think.
Nevertheless, we can think about using `Integer` keys for the graph.
It takes 3ms in this way.
(We can also use `List<TreeNode>[] graph`, but the map is fine I think.)

```java
class Solution {

  Map<Integer, List<TreeNode>> graph = new HashMap<>();
 

  public int findClosestLeaf(TreeNode root, int k) {
    if (root.left == null && root.right == null) {
      return k;
    }
 
    createGraphWithDFS(root, null);

    return shortestPathBFS(root, k);
  }

  private int shortestPathBFS(TreeNode root, int k) {
    Deque<Integer> queue = new ArrayDeque<>();
    boolean[] seen = new boolean[1001];

    queue.offer(k);
    seen[k] = true;

    while (!queue.isEmpty()) {
      Integer node = queue.poll();
      
        List<TreeNode> children = graph.get(node);  
        if (node != root.val && children.size() <= 1) {
          return node;
        }
        for (TreeNode adjacentNode : children) {
          if ( !seen[adjacentNode.val]) {
            seen[adjacentNode.val] = true;
            queue.add(adjacentNode.val);
          }
        }
      
    }

   return -1;
  }

  public void createGraphWithDFS(TreeNode node, TreeNode parent) {
    if (node != null) {
    
      graph.put(node.val, new ArrayList<>());
      if (parent != null) {
        graph.get(node.val).add(parent);
        graph.get(parent.val).add(node);
      }

      createGraphWithDFS(node.left, node);
      createGraphWithDFS(node.right, node);
    }
  }
}
```

Reply:
We can do such thins here because the Node values are unique. That was also the case with the boolean array `seen`.

Reply:
Changes added.